### PR TITLE
Add gazebo_ros::QoS class

### PR DESCRIPTION
**Pull request by jacobperron**
Connects to https://github.com/ros-simulation/gazebo_ros_pkgs/issues/1079

This PR adds a new class that parses SDF for ROS 2 QoS settings that plugins can use when creating publishers and subscriptions.

Here's an example of valid SDF (from one of the tests):

```xml
<?xml version='1.0' ?>
<sdf version='1.6'>
  <world name='default'>
    <plugin name='node_1' filename='libnode_name.so'>
      <ros>
        <qos>
          <topic name="foo">
            <publisher>
              <reliability>best_effort</reliability>
              <durability>transient_local</durability>
              <history>keep_all</history>
              <deadline>123</deadline> <!-- milliseconds -->
              <lifespan>42</lifespan> <!-- milliseconds -->
              <liveliness>manual_by_topic</liveliness>
              <liveliness_lease_duration>1000</liveliness_lease_duration> <!-- milliseconds -->
            </publisher>
            <subscription>
              <reliability>reliable</reliability>
              <durability>volatile</durability>
              <history depth="3">keep_last</history>
            </subscription>
          </topic>
        </qos>
      </ros>
    </plugin>
  </world>
</sdf>
```

The new `<qos>` tag can contain zero or more `<topic>` tags. Each topic tag must have a `name` attribute.

The implementation will look for optional `<publisher>` and `<subscription>` tags, which can contain any number of tags corresponding to all of the QoS settings supported by ROS 2.

A description of QoS settings can be found here:

- http://design.ros2.org/articles/qos.html
- http://design.ros2.org/articles/qos_deadline_liveliness_lifespan.html

One limitation of the current approach is that **all** subscriptions (or publishers) for a given topic will use the same QoS overrides. I don't think there's a way to differentiate them in ROS 2, but I also don't think this will be an issue for any of the plugins in this repository.

The choice to use milliseconds for all duration values is arbitrary. If there is a more preferable unit, it's easy to change.

### CODE REVIEW

**Comment by jacobperron** [comment]
Here is a follow-up PR that makes use of the new `gazebo_ros::QoS` object in gazebo_plugins: https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1092

File: gazebo_ros/include/gazebo_ros/node.hpp
Comment (ivanpauno) [inline gazebo_ros/include/gazebo_ros/node.hpp]: can be `const &` ?

File: gazebo_ros/test/test_qos.cpp
Comment (ivanpauno) [inline gazebo_ros/test/test_qos.cpp:38]: I would use https://github.com/ros2/rclcpp/blob/e2dbc5d5d5fbfc9a4b929e0eb73c4d150082e08e/rclcpp/include/rclcpp/qos.hpp#L43 with `rmw_qos_profile_default`, just in case the default profile gets updated.

File: gazebo_ros/include/gazebo_ros/qos.hpp
Comment (ivanpauno) [inline gazebo_ros/include/gazebo_ros/qos.hpp]: why is this needed?
why not just creating a `rclccp::QoS` object from the SDF?

File: gazebo_ros/src/qos.cpp
Comment (ivanpauno) [inline gazebo_ros/src/qos.cpp]: nit: instead of multiple `if-else` blocks, we could use a map. IMHO, it's easier to read and extend.

File: gazebo_ros/src/qos.cpp
Comment (ivanpauno) [inline gazebo_ros/src/qos.cpp]: This makes me remember that I should have removed that option :joy:

File: gazebo_ros/src/qos.cpp
Comment (ivanpauno) [inline gazebo_ros/src/qos.cpp:299]: `lifespan` is only valid for a `Publisher`, and not for a `Subscription`.
This is actually a problem in `rmw` and `rclcpp`, so I don't mind.

File: gazebo_ros/include/gazebo_ros/node.hpp
Comment (ivanpauno) [inline gazebo_ros/include/gazebo_ros/node.hpp]: Is this parameter needed here?
I think that `static SharedPtr Get(sdf::ElementPtr _sdf);` can directly do all the work, without forwarding a parameter to this function.

File: gazebo_ros/src/qos.cpp
Comment (ivanpauno) [inline gazebo_ros/src/qos.cpp:310]: We will need to make use of this function and `get_subscription_qos` in all plugins, right?

File: gazebo_ros/test/test_qos.cpp
Comment (ivanpauno) [inline gazebo_ros/test/test_qos.cpp:50]: nit: I would indent (using spaces in the strings) in these in line `xml` examples.
IMHO, it's more readable with indentation.

**Review by ivanpauno — COMMENTED** [review]
Great work @jacobperron !!

I've left some minor comments

File: gazebo_ros/src/qos.cpp
Comment (jacobperron) [reply gazebo_ros/src/qos.cpp:310]: That's right. See the connected PR for usage: https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1092

File: gazebo_ros/test/test_qos.cpp
Comment (jacobperron) [reply gazebo_ros/test/test_qos.cpp:50]: Yeah, I'll try doing that. I was just following the style of the other tests in `gazebo_ros`.

File: gazebo_ros/include/gazebo_ros/qos.hpp
Comment (jacobperron) [reply gazebo_ros/include/gazebo_ros/qos.hpp]: The problem is we are overlaying the QoS settings from the SDF onto the default provided by the plugin. In the current design, the default QoS are provided when the plugin calls `get_*_qos()`. I.e. we don't know the default QoS when parsing the SDF for overrides. The other alternative I can think of would be to parse the SDF each time `get_*_qos()` is called, but that seems redundant.

File: gazebo_ros/src/qos.cpp
Comment (jacobperron) [reply gazebo_ros/src/qos.cpp:299]: Yeah, unless the behaviour changes in `rclcpp`, I don't see what else we can do here. We could log a warning, but maybe that's something best left to `rclcpp`.

File: gazebo_ros/src/qos.cpp
Comment (jacobperron) [reply gazebo_ros/src/qos.cpp]: I don't know why, but should we not include it here then?

File: gazebo_ros/include/gazebo_ros/node.hpp
Comment (jacobperron) [reply gazebo_ros/include/gazebo_ros/node.hpp]: Yep, 34b862d

File: gazebo_ros/test/test_qos.cpp
Comment (jacobperron) [reply gazebo_ros/test/test_qos.cpp:50]: Done: 4ca784e

File: gazebo_ros/test/test_qos.cpp
Comment (jacobperron) [reply gazebo_ros/test/test_qos.cpp:38]: I just explicitly set the default QoS being tested with instead: d85fbd6

File: gazebo_ros/include/gazebo_ros/node.hpp
Comment (jacobperron) [reply gazebo_ros/include/gazebo_ros/node.hpp]: You're right. Removed the parameter in da5b314

File: gazebo_ros/src/qos.cpp
Comment (jacobperron) [reply gazebo_ros/src/qos.cpp]: Good idea, see 728bb0e

File: gazebo_ros/src/qos.cpp
Comment (ivanpauno) [reply gazebo_ros/src/qos.cpp]: That feature depended on participants being mapped to nodes, and there was agreement to remove it.
I will create a couple of PRs to actually remove it, but it can already be removed from here.

File: gazebo_ros/include/gazebo_ros/qos.hpp
Comment (ivanpauno) [reply gazebo_ros/include/gazebo_ros/qos.hpp]: Oh, I see. The current solution looks fine :+1:

File: gazebo_ros/src/qos.cpp
Comment (ivanpauno) [inline gazebo_ros/src/qos.cpp]: nit^2: prefer `std::unordered_map`

File: gazebo_ros/src/qos.cpp
Comment (ivanpauno) [inline gazebo_ros/src/qos.cpp]: nit^N: prefer `std::ostringstream`.

**Review by ivanpauno — APPROVED** [review]

**Comment by jacobperron** [comment]
I don't think the current implementation will work with remapped topic names. I'm going to look into it and add some tests.

**Comment by ivanpauno** [comment]
> I don't think the current implementation will work with remapped topic names. I'm going to look into it and add some tests.

Great point.
I was going to ask that question in my first review, and then I completely forgot :joy:.

IMO, remapping should be done before querying the qos profile.
i.e. I expect a `SDF` with both remapping rules and QoS settings to look like:

```
<ros>
     <remapping>topic:=my_topic</remapping>
</ros>
 <qos>
    <topic name='my_topic'>
        ...
    </topic>
</qos>
```

For that, I think we need to apply remapping rules/topic name expansion in `get_*_qos()`.

P.S.: Shouldn't `qos` tag be a child of `ros` tag?

**Comment by jacobperron** [comment]
I agree with your assessment. I'll take a look at implementing it now.

> P.S.: Shouldn't qos tag be a child of ros tag?

Yes, when the SDF is parsed here in `gazebo_ros::Node`:

https://github.com/ros-simulation/gazebo_ros_pkgs/blob/728bb0e05023d78b8e9e9d663ace58d08134f1aa/gazebo_ros/src/node.cpp#L105

The `sdf` is a `<ros>` element.

I'll update the example in this PR description.

For unit testing `gazebo_ros::QoS`, it's not strictly required.

File: gazebo_ros/src/qos.cpp
Comment (jacobperron) [reply gazebo_ros/src/qos.cpp]: updated dc9546f

File: gazebo_ros/src/qos.cpp
Comment (jacobperron) [reply gazebo_ros/src/qos.cpp]: updated a7eaa63

File: gazebo_ros/src/qos.cpp
Comment (jacobperron) [reply gazebo_ros/src/qos.cpp]: Removed in b52f8c2

**Comment by jacobperron** [comment]
@ivanpauno Here's my attempt at supporting remapping: 9e753f5

I don't believe there's a way to get the topic remappings from `rclcpp`, so I'm making the necessary `rcl` function calls.

**Comment by ivanpauno** [comment]
@jacobperron By mistake I did my comments in commit https://github.com/ros-simulation/gazebo_ros_pkgs/commit/9e753f5e6f306e759e2864efc2d4814b1059564e, if you open it you can see the comments in their context.

It would be nice to add a test were a namespace is specified in the `sdf`.
So, it's clear how both expansion and remapping interact with this new feature.

**Review by ivanpauno — APPROVED** [review]
LGTM!!

File: gazebo_ros/include/gazebo_ros/qos.hpp
Comment (scpeters) [inline gazebo_ros/include/gazebo_ros/qos.hpp]: @chapulina I'm not sure if we're using the PIMPL pattern in `gazebo_ros_pkgs` yet, but I'm wondering if this new class is a good candidate for it, since it has lots of private member variables

**Comment by jacobperron** [comment]
> It would be nice to add a test were a namespace is specified in the sdf.

I've updated the test to include a node namespace: 19469da

**Review by ivanpauno — APPROVED** [review]

**Comment by ivanpauno** [comment]
@jacobperron @scpeters is this ready to go in?

**Comment by jacobperron** [comment]
@ivanpauno I think the point about considering to use a pimpl pattern is something we should consider (https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1091#discussion_r424761989).

File: gazebo_ros/include/gazebo_ros/node.hpp
Comment (ivanpauno) [reply gazebo_ros/include/gazebo_ros/node.hpp]: I was referring to the returned object :joy:.
If the returned value is a `const &`, it's also good to keep the ref qualifier for the method.

Comment (jacobperron) [reply gazebo_ros/include/gazebo_ros/node.hpp]: Ooops, that's what I meant to do, but wasn't thinking :upside_down_face:

Comment (jacobperron) [reply gazebo_ros/include/gazebo_ros/node.hpp]: Fixed in 5bcfe93

**Comment by jacobperron** [comment]
I rebased on `ros2` to get changed needed for compilation on Ubuntu Focal (https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1093).

File: gazebo_ros/include/gazebo_ros/qos.hpp
Comment (chapulina) [reply gazebo_ros/include/gazebo_ros/qos.hpp]: Thank you for catching this, @scpeters . Indeed we're using PIMPL for the ROS 2 version of `gazebo_ros_pkgs`, as described on the [Contribution guide](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-contribution-guide#in-the-code).

**Comment by ivanpauno** [comment]
@scpeters @jacobperron I finally figured out the remaining build problems before closing.
I've introduced some linter failures that I will fix tomorrow, but you can already take a look to the last two commits (https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1091/commits/1eb881aca037fd80868f54df31bd89ba4e838332, https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1091/commits/3b8d0f535047883bad91d1d584e787add600fc26).

**Comment by jacobperron** [comment]
The latest changes LGTM :+1:

File: gazebo_ros/src/node.cpp
Comment (chapulina) [inline gazebo_ros/src/node.cpp:41]: Is this change needed?

File: gazebo_ros/test/test_qos.cpp
Comment (chapulina) [inline gazebo_ros/test/test_qos.cpp]: I know both `<plugin><qos>` and `<plugin><ros><qos>` are supported, but I guess the latter is the recommended one (based on documentation). How about using that on most tests? Just so people looking have a good example to follow.

**Review by chapulina — COMMENTED** [review]
I haven't tried running it, but I skimmed through the code and it looks good to me.

Besides those 2 comments, I'd also ask if it would be possible to use the new `<qos>` elements in one of the example worlds [here](https://github.com/ros-simulation/gazebo_ros_pkgs/tree/ros2/gazebo_ros/worlds) or [here](https://github.com/ros-simulation/gazebo_ros_pkgs/tree/ros2/gazebo_plugins/worlds), or create a new example world to demonstrate the feature.

File: gazebo_ros/src/node.cpp
Comment (ivanpauno) [reply gazebo_ros/src/node.cpp:41]: Valid namespaces should be always absolute (contrary to names, which can be relative/private/absolute).

I guess @jacobperron did that change because the namespace is being passed to the new QoS class, and some of the API it uses may be doing namespace validation.

@jacobperron can you confirm if that was the case?

Comment (chapulina) [reply gazebo_ros/src/node.cpp:41]: Got it. Since we started going down this road on this PR, do you think we should also do some checking on line 58 in case the user passes an invalid namespace? Or maybe we prepend the `/` for them in case it's missing?

File: gazebo_ros/test/test_qos.cpp
Comment (ivanpauno) [reply gazebo_ros/test/test_qos.cpp]: Done in https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1091/commits/7a5188abb9177bb28ef644d1a2def14fa7193cfa

**Comment by ivanpauno** [comment]
> Besides those 2 comments

I've addressed/replied to the commets.

> I'd also ask if it would be possible to use the new <qos> elements in one of the example worlds here or here, or create a new example world to demonstrate the feature.

I would prefer to do that in a follow up if you don't mind.
Feel free to create a ticket and assign it to me.

File: gazebo_ros/src/node.cpp
Comment (jacobperron) [reply gazebo_ros/src/node.cpp:41]: @ivanpauno is exactly right. Without this change, I was seeing runtime errors from some `rcl` / `rclcpp` API calls (though I can't recall exactly which ones).

I guess if a user passes an invalid namespace, they will see similar runtime errors. I think the validation and error message from the underlying layers is enough (it also is nice to keep it encapsulated there in case the naming requirement changes for some reason).

Comment (chapulina) [reply gazebo_ros/src/node.cpp:41]: Got it, makes sense. I'm just wondering why we haven't seen problems so far with this default value, and whether this change is backwards compatible. But you peeps know best about these details :wink:

Comment (jacobperron) [reply gazebo_ros/src/node.cpp:41]: It's a good question. It looks like the [node initialization logic handles an empty string specifically](https://github.com/ros2/rcl/blob/85ef95514d6ee47bb8285009add84e956595c018/rcl/src/rcl/node.c#L173-L176), but the function for [expanding topic names](https://github.com/ros2/rcl/blob/85ef95514d6ee47bb8285009add84e956595c018/rcl/src/rcl/expand_topic_name.c#L44) does not. It's possible the latter function could be updated in the future to handle this case, but at the end of the day and empty string will resolve to `"/"`.

**Comment by ivanpauno** [comment]
@chapulina is this ready to go in?

**Comment by chapulina** [comment]
> @chapulina is this ready to go in?

I haven't tried it out but looking at the code I don't see anything weird. Could you peeps just run CI before merging? Instructions here: https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-CI#next-turtle

**Comment by ivanpauno** [comment]
> I haven't tried it out but looking at the code I don't see anything weird. Could you peeps just run CI before merging? Instructions here: https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-CI#next-turtle

Thanks for the instructions! I didn't know how to run CI here.

CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10868/)

**Review by chapulina — APPROVED** [review]
Thanks for triggering CI. I see all QoS tests passed. I don't think the flake8 failures are this PR's fault.

All other tests failing with:

`[Err] [Master.cc:95] EXCEPTION: Unable to start server[bind: Address already in use]. There is probably another Gazebo process running.`

are being caused by some earlier test not being properly terminated and leaving Gazebo running. I think it's because of `test_entity_spawner.test.py`, see the testing adventure here: https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1033#issuecomment-577363549. Again, not this PR's fault.

**Comment by ivanpauno** [comment]
> Thanks for triggering CI. I see all QoS tests passed. I don't think the flake8 failures are this PR's fault.

I saw the flake8 failures locally, and they are also reproducible in `ros2` branch.

> All other tests failing with:
>
> [Err] [Master.cc:95] EXCEPTION: Unable to start server[bind: Address already in use]. There is probably another Gazebo process running.
> 
> are being caused by some earlier test not being properly terminated and leaving Gazebo running. > I think it's because of test_entity_spawner.test.py, see the testing adventure here: #1033 (comment). Again, not this PR's fault.

I experienced that locally too :joy:.
Killing the gazebo server running in the background before re-running the tests solved the issue (we should probably have a script in CI machines to kill it).

----

@chapulina I don't have write access here, please merge if it's ready.
Thanks!!

**Comment by jacobperron** [comment]
I can rebase the connected PR.

**Comment by chapulina** [comment]
> we should probably have a script in CI machines to kill it

Yeah I think this could be built into the test cases themselves to kill any running `gzserver`s in between tests. The downside is if someone is running another simulation on the same machine and that gets killed by mistake :sweat_smile:

File: gazebo_ros/src/node.cpp
Comment (chapulina) [reply gazebo_ros/src/node.cpp:41]: I think this may have messed something up :thinking: https://github.com/ros-simulation/gazebo_ros_pkgs/issues/1116

Comment (ivanpauno) [reply gazebo_ros/src/node.cpp:41]: Doesn't seem related to me.
That default value is being immediately overridden by the value in the SDF if it exists https://github.com/ros-simulation/gazebo_ros_pkgs/blob/57732fe150efcb324e905f0f68537888c1eceb30/gazebo_ros/src/node.cpp#L57-L59.

Comment (ivanpauno) [reply gazebo_ros/src/node.cpp:41]: Maybe, it's related to this https://github.com/ros-simulation/gazebo_ros_pkgs/blob/57732fe150efcb324e905f0f68537888c1eceb30/gazebo_ros/src/gazebo_ros_factory.cpp#L231-L235
https://github.com/ros-simulation/gazebo_ros_pkgs/blob/57732fe150efcb324e905f0f68537888c1eceb30/gazebo_ros/src/gazebo_ros_factory.cpp#L354-L355
?

According to grep that's the only place where we set the `namespace` SDF tag.

Comment (chapulina) [reply gazebo_ros/src/node.cpp:41]: Nice catch! @ahcorde , are you spawning your robot through the factory? In my case I was.

Comment (ahcorde) [reply gazebo_ros/src/node.cpp:41]: https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1117

**Comment by srushtibobade** [comment]
@jacobperron 
Thanks for all the good work!!
 where should this XML file be located/ or added to the project?

**Comment by ivanpauno** [comment]
this can just be added in a normal gazebo sdf world file, the new qos parameters can be applied to gazebo_ros_pkgs plugins as shown in the first post https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1091#issue-608735085

**Pull request merged**
